services/core/shared/core/delegates.js

```diff
@@ -119,8 +132,111 @@ const getTotalNumberOfDelegates = async (params = {}) => {
 	return relevantDelegates.length;
 };
 
+const computeDelegateRankAndStatus = async () => {
+	const db = await pouchdb(config.db.collections.delegates.name);
+	const sdkVersion = Number(coreApi.getSDKVersion().split('sdk_v')[1]);
+	const numActiveForgers = 101;
+
+	const delegateStatus = {
+		ACTIVE: 'active',
+		STANDBY: 'standby',
+		BANNED: 'banned',
+		PUNISHED: 'punished',
+		NON_ELIGIBLE: 'non-eligible',
+	};
+
+	const allDelegates = await db.findAll();
+	const lastestBlock = await getBlocks({ limit: 1 });
+	const allNextForgersAddressList = nextForgers.map(forger => forger.account.address);
+	const activeNextForgersList = allNextForgersAddressList.slice(0, numActiveForgers);
+
+	const verifyIfPunished = delegate => {
+		const isPunished = delegate.pomHeights
+			.some(pomHeight => pomHeight.start <= lastestBlock.height
+				&& lastestBlock.height <= pomHeight.end);
+		return isPunished;
+	};
+
+	if (sdkVersion >= 4) allDelegates.sort(delegateComparator);
+	allDelegates.map((delegate, index) => {
+		if (sdkVersion < 4) {
+			if (activeNextForgersList.includes(delegate.account.address)) {
+				delegate.status = delegateStatus.ACTIVE;
+			} else delegate.status = delegateStatus.STANDBY;
+		} else {
+			delegate.rank = index + 1;
+			if (!delegate.isDelegate) delegate.status = delegateStatus.NON_ELIGIBLE;
+			else if (delegate.isBanned) delegate.status = delegateStatus.BANNED;
+			else if (verifyIfPunished(delegate)) delegate.status = delegateStatus.PUNISHED;
+			else if (activeNextForgersList.includes(delegate.account.address)) {
+				delegate.status = delegateStatus.ACTIVE;
+			} else delegate.status = delegateStatus.STANDBY;
+		}
+
+		return delegate;
+	});
+
+	if (allDelegates.length) await db.writeBatch(allDelegates);
+};
+
+const getNextForgers = async params => {
+	let forgers = {
+		data: [],
+		meta: {},
+		links: {},
+	};
+
+	try {
+		if (nextForgers.length) {
+			const offset = params.offset || 0;
+			const limit = params.limit || 10;
+
+			forgers.data = nextForgers.slice(offset, offset + limit);
+
+			forgers.meta.count = forgers.data.length;
+			forgers.meta.offset = offset;
+			forgers.meta.total = nextForgers.length;
+		} else throw new Error('Request Next Forgers data from Lisk Core');
+	} catch (err) {
+		logger.debug(err.message);
+
+		forgers = await coreApi.getNextForgers(params);
+		forgers.data = await BluebirdPromise.map(
+			forgers.data,
+			async forger => {
+				const forgerDelegateInfo = await getDelegates({ address: forger.address });
+				return forgerDelegateInfo.data[0];
+			},
+			{ concurrency: forgers.data.length });
+		forgers.data.sort(delegateComparator);
+	}
+	return forgers;
```

Comment:
With that logic in mind, is the logic written in the controller still needed? I mean `methods/controllers/delegates.js`

Reply:
`ensureDelegatesArr` inside of `getNextForgers` within the controller is only necessary during the cold start-up of Lisk Service so that the status of the delegates is properly set because as the initial jobs responsible to set the status would still be processing the complete information. Also, since `await CoreService.getDelegates({ address: forger.account.address });` is a guaranteed hit in the in-mem DB cache thereafter, the subsequent calls should be inexpensive.

Reply:
Removed the logic from the controller and made necessary changes at the abstraction layer.